### PR TITLE
Implementing cached dataset querying

### PR DESCRIPTION
**Pull request by Shalelol**
# Changes
- Implementing a way to query the new [cached datasets](https://keen.io/docs/api/#cached-datasets) API. 
- Adding the user of a [Snake Case Contract Resolver](https://gist.github.com/crallen/9238178) for NewtonSoft.Json because the use of the JsonPropertyAttribute can get a bit unwieldy once the data structures get larger and more complex, as they can with the cached datasets.

# Comments
- I am moving towards strongly typed models, leveraging the convenience of the contract resolver, and trying to work with JToken as infrequently as possible, please let me know if anyone has any issues with this.

- I'm trying to adhere to the naming and code styling convention as much as possible, please do let me know if I stray from that too much.

- I'm quite inexperienced with unit testing, particularly with NUnit, so would love to hear any feedback or advice regarding that.

- I plan to finish work on this pull request mid to late August this year (2017).

# Issues addressed
- #92 - Delete Cached Dataset
- #91 - List Cached Dataset Definitions
- #90 - Retrieve Cached Dataset Results 
- #89 - Get Cached Dataset Definition
- #88 - Create Cached Dataset

### CODE REVIEW

**Comment by masojus** [comment]
Thanks @Shalelol I will take a look! Meanwhile could you specify which Issues this resolves? Thanks!

**Comment by Shalelol** [comment]
Hey @masojus I've updated my initial comment to include the issues addressed so far. I also may look into addressing more of the issues as my requirements grow.

**Comment by masojus** [comment]
Thank you @Shalelol! And don't worry about the AppVeyor failure. I imagine it's an expected failure I need to deal with but all the tests look to have passed. I presume manual testing went well? Or are there any major corner cases to watch out for?

**Comment by Shalelol** [comment]
@masojus So far all manual testing is working perfectly. I will inform you if any corner cases pop up.

I'm also still yet to write a unit test for `GetDatasetDefinition` but will be doing so shortly.

File: Keen/ContractResolvers/SnakeCaseContractResolver.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
+﻿namespace Keen.Core.ContractResolvers
+{
+    using Newtonsoft.Json.Serialization;
+
+    /* 
+     * Found on GithubGist https://gist.github.com/crallen/9238178
+     * 
+     * Credit to Chris Allen https://github.com/crallen 
+     */
+    public class SnakeCaseContractResolver : DefaultContractResolver
```
Comment (masojus) [inline Keen/ContractResolvers/SnakeCaseContractResolver.cs]: Is this necessary because the version of Json.NET we're using doesn't yet have the `SnakeCaseNamingStrategy` or is it the case that even that strategy does not satisfy your needs? How does this `SnakeCaseContractResolver` handle `camelCase` or `PascalCase` or other casing like `kebab-case`, and how does that compare to Json.NET's own `SnakeCaseNamingStrategy`? How does this treat names that have a `JsonPropertyAttribute` explicitly specifying the name?

http://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Serialization_SnakeCaseNamingStrategy.htm
http://www.phrohdoh.com/blog/2016/09/02/newtonsoft-json-custom-property-names/

File: Keen/Dataset/Datasets.cs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    public interface IDataset
+    {
+        Task<JObject> Results(string datasetName, string indexBy, string timeframe);
+        Task<DatasetDefinition> Definition(string datasetName);
+    }
+
+    internal class Datasets : IDataset
+    {
+        private readonly IKeenHttpClient _keenHttpClient;
+        private readonly string _cachedDatasetReleativeUrl;
```
Comment (masojus) [inline Keen/Dataset/Datasets.cs]: typo: `_cachedDatasetReleativeUrl`->`_cachedDatasetRelativeUrl`

File: Keen/KeenClient.cs [hunk 3]
```diff
@@ -970,5 +975,39 @@ public JObject Query(string queryName, Dictionary<string, string> parms)
                 throw ex.TryUnwrap();
             }
         }
+
+        public async Task<JToken> QueryDatasetAsync(string datasetName, string indexBy, string timeframe)
```
Comment (masojus) [inline Keen/KeenClient.cs]: We'll want doc comments on these public methods.

File: Keen/Dataset/Datasets.cs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+namespace Keen.Core.Dataset
+{
+    using System.Threading.Tasks;
+    using ContractResolvers;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Linq;
+    using Query;
+
+    public interface IDataset
```
Comment (masojus) [inline Keen/Dataset/Datasets.cs]: I would tend to prefer interfaces in their own file, and public API surface (types, fields, methods, etc.) should have doc comments.

File: Keen/ContractResolvers/SnakeCaseContractResolver.cs (hunk 1 above)
Comment (Shalelol) [inline Keen/ContractResolvers/SnakeCaseContractResolver.cs]: I've used the Contract Resolver rather than Json.NET because all the attributes defined on all the existing models are using the JsonPropertyAttribute from Newtonsoft.Json. Which will only be respected when using the `JsonConvert.SerializeObject()` and `JsonConvert.DeserializeObject()` functions.

The `SnakeCaseContractResolver` doesnt handle `camelCase`, `PascalCase` or `kebab-case` but custom resolvers may be defined for those. The resolver is defined at serialization/deserialization time by passing it as an option to JsonConvert like so, rather than in a global setting:
```
JsonConvert.DeserializeObject<DatasetDefinition>(responseString, new JsonSerializerSettings { ContractResolver = new SnakeCaseContractResolver()});
```

This, and indeed any ContractResolver extending Newtonsoft.Json's `DefaultContractResolver` will respect all JsonPropertyAttributes

File: Keen/Dataset/Dataset.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Keen.Core.Dataset
+{
+    using Query;
```
Comment (masojus) [inline Keen/Dataset/Dataset.cs]: Reminds me we need to start creating some guidelines around things like `using` inside and outside of namespaces and configure some stylecop rulesets: https://github.com/DotNetAnalyzers/StyleCopAnalyzers

**Comment by Shalelol** [comment]
Hey @masojus, I just want to apologize for the messy state of some of the code. I just realized that I didn't really communicate how WIP this Pull Request is. I had/have every intention of coming back and cleaning everything up, things such as xml comments, moving files out and scoping usings to match the rest of the project. This is still very much an in progress project at the moment as we move all of our querying to cached querying.

I was just putting the WIP PR up nice and early so that it's on your radar =]

File: Keen/Dataset/Datasets.cs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+            KeenUtil.CheckApiErrorCode(response);
+
+            if (!responseMessage.IsSuccessStatusCode)
+            {
+                throw new KeenException("Request failed with status: " +
+                                        responseMessage.StatusCode);
+            }
+
+            return JsonConvert.DeserializeObject<DatasetDefinition>(responseString, new JsonSerializerSettings { ContractResolver = new SnakeCaseContractResolver()});
```
Comment (masojus) [inline Keen/Dataset/Datasets.cs]: I'll have to think about this and ponder the answers to the questions in my comment above about `SnakeCaseContractResolver` and what it means to always use this resolver for all Dataset definitions.

**Comment by masojus** [comment]
@Shalelol No worries. I'm just skimming it anyway, so I'll take a closer look at it later when you've tidied up a bit and think it's ready for prime time. 👍  Thanks!

**Comment by tbarn** [comment]
@Shalelol just wanted to say, thanks for working on this! We really appreciate it.

**Comment by masojus** [comment]
We don't have the Travis config merged into master yet, so I wouldn't worry too much about that one.

**Comment by Shalelol** [comment]
Just updated the opening comment with issues. I plan to implement #88, #91 and #92 for completeness.

**Comment by Shalelol** [comment]
Coveralls likes to point out when you've been bad, but not when you're trying to improve the coverage 😞.

**Comment by baumatron** [comment]
I think It should hand out gold stars and/or free beers for increased coverage ⭐️ 🍻.

**Comment by Shalelol** [comment]
Hey @masojus. I've finished implementing all the Cached Dataset issues, and I've brought the coverage of `Keen.Core.Dataset` up to 91%. I've also gone through and added XML comments to everything that is public facing.

Could you please give this another review when you have time? 

Thanks 😃.

**Comment by masojus** [comment]
Thank you! We'll get some eyes on this as soon as possible.

**Comment by masojus** [comment]
Sorry for the delay here. We're going to be merging the dotnetsummer branch shortly and will address this after that so as to also port it to .NET Standard 2.0 while we're at it.

**Comment by masojus** [comment]
Hello @Shalelol, we've now merged the big chunk of work we wanted to get into master before taking this PR on. Could you merge from master and update the PR? If you need help with that we can pitch in, but then we could work on getting this PR in. Thanks!

**Comment by masojus** [comment]
I created branch `jm_UpdateCachedDatasetsBranch` to get this up to date with `master` and continue the conversation. I'll probably close this PR and we can finish polishing that off in the new PR.

**Comment by masojus** [comment]
See PR #128 for a continuation of this PR.
